File: 4 kyu/Twice linear/solution.php
function dblLinear($n) {
    $h = 1; $cnt = 0; $q2 = array(); $q3 = array();
    while (true) {
        if ($cnt >= $n) {
            return $h;
        }
        array_push($q2, 2 * $h + 1);
        array_push($q3, 3 * $h + 1);
        $h = min($q2[0], $q3[0]);
        if ($h === $q2[0]) {
            $h = array_shift($q2);
        }
        if ($h === $q3[0]) {
            $h = array_shift($q3);
        }
        $cnt++;
    }
}

__________________________________________________
function dblLinear(int $n): int {
  foreach (genU() as $i => $v) {
    if ($i == $n) { 
      return $v;
    }
  } 
}

function genU(): Generator {
  $u = new SplMinHeap();

  $x = 1;
  $u->insert($x);

  while (true) { 
    $u->insert(2 * $x + 1);
    $u->insert(3 * $x + 1);

    yield $x;

    do {
      $u->next();
      $newX = $u->current();
    } while ($newX == $x);

    $x = $newX;
  }
}

__________________________________________________
function numberIsInSequence($n) {
    if ($n < 1 || $n != intval($n)) {
        return false;
    } else if ($n === 1) {
        return true;
    } else if ($n % 2 === 0) {
        return numberIsInSequence(($n-1)/3);
    } else {
        return numberIsInSequence(($n-1)/2) || numberIsInSequence(($n-1)/3);
    }
}

function dblLinear($n) {
    $position = 1;
    $i = 1;
    while($position <= $n) {
        $i++;
        if (numberIsInSequence($i))
            $position++;
    }
    return $i;
}

File: 4 kyu/Twice linear/solution.swift
func dblLinear(_ n: Int) -> Int {
    var array = [1]
    
    var x = 0
    var y = 0
    while array.count <= n {
        let xx = 2 * array[x] + 1
        let yy = 3 * array[y] + 1
        
        if xx < yy {
            array.append(xx)
            x += 1
        } else if xx > yy {
            array.append(yy)
            y += 1
        } else {
            array.append(yy)
            x += 1
            y += 1
        }
    }
    
    return array[n]
}

__________________________________________________
func dblLinear(_ n: Int) -> Int {
    var u = [1], x = 0, y = 0
    (0..<n).forEach { _ in
        let nextX = 2 * u[x] + 1, nextY = 3 * u[y] + 1
        if nextX <= nextY {
            u.append(nextX); x += 1
            if nextY == nextX {y += 1}
        } else {
            u.append(nextY); y += 1
        }
    }
    return u[n]
}

__________________________________________________
func dblLinear(_ n: Int) -> Int {
  var u = Array(repeating: 0, count: n + 1)
  u[0] = 1
  var x = 0
  var y = 0
  var i = 1
  while (i <= n) {
      u[i] = min(2 * u[x] + 1, 3 * u[y] + 1)
      if (u[i] == 2 * u[x] + 1) {
          x += 1
      }
      if (u[i] == 3 * u[y] + 1) {
          y += 1
      }
      i += 1
  }
  return u[n]
}
